File: py/openaxis/src/openaxis/navigation_diagnostics.py
"""Passive navigation evidence and renderer-independent presentation.

No application calls, timers, graphics dependencies or protocol messages. Call on
the session thread; render the returned detached presentation on the UI thread.
"""
from collections import deque
from copy import deepcopy
from dataclasses import dataclass, field
from itertools import product
import math
import time

from .navigation import UNAVAILABLE, _wire_fact_value
from .geometry import Quat, Vec3
from .diagnostics import format_event
from ._session import compare, compare_object


COLORS = {
    'text': (245,245,245), 'missing': (255,130,130), 'skipped': (165,165,165),
    'pass': (80,255,110), 'selection': (255,150,40), 'model': (40,210,255),
    'target': (255,70,220), 'cursor': (255,235,40), 'center': (100,170,255),
    'object': (255,70,220), 'sketch': (190,120,255), 'axis_x': (255,60,60), 'axis_y': (60,255,60), 'axis_z': (60,130,255),
    'ray': (175,175,175), 'correction': (255,210,40),
}


@dataclass(frozen=True)
class DiagnosticLine:
    text: str
    tone: str = 'text'


@dataclass(frozen=True)
class DiagnosticSegment:
    start: tuple
    end: tuple
    tone: str
    width: float = 1
    opacity: float = 1  # Candidate evidence is subdued; this does not imply server use.


@dataclass(frozen=True)
class DiagnosticMarker:
    label: str
    point: tuple  # Application-defined local renderer coordinates.
    tone: str


@dataclass(frozen=True)
class DiagnosticPresentation:
    context: object
    lines: tuple = ()
    segments: tuple = ()  # Application world coordinates, projected at draw time.
    markers: tuple = ()
    revision: int = 0
    expires_at: float | None = None


@dataclass
class QueryEvidence:
    request_id: int
    gesture_id: int | None
    values: tuple
    first: tuple
    context: object = None
    facts: dict = field(default_factory=dict)
    picks: dict = field(default_factory=dict)
    selected: str | None = None
    duration_ms: float = 0
    error: str | None = None
    complete: bool = False


def _tone(name):
    if name.startswith('pick.'):
        return 'cursor' if 'cursor' in name else 'center'
    return {'selection.bounds': 'selection', 'model.bounds': 'model',
            'object.bounds': 'object', 'object.pose': 'object', 'scene.cursor': 'target',
            'sketch.plane': 'sketch', 'camera.view_target': 'target'}.get(name, 'text')


def _point(value):
    if isinstance(value, (list, tuple)) and len(value) == 3:
        try:
            result = tuple(float(x) for x in value)
            return result if all(math.isfinite(x) for x in result) else None
        except (TypeError, ValueError):
            pass
    return None


def _bounds(value):
    if isinstance(value, dict):
        low, high = _point(value.get('min')), _point(value.get('max'))
        if low and high and all(a <= b for a,b in zip(low,high)):
            return low,high
    return None


def _box(bounds, tone):
    corners = list(product(*zip(*bounds)))
    for i, start in enumerate(corners):
        for bit in (1,2,4):
            if not i & bit:
                yield DiagnosticSegment(start,corners[i | bit],tone,1,.35)


def _cross(point, radius, tone):
    for axis in range(3):
        a,b = list(point),list(point)
        a[axis] -= radius
        b[axis] += radius
        yield DiagnosticSegment(tuple(a),tuple(b),tone,1,.35)


class NavigationDiagnostics:
    """Optional session consumer. Logging remains independent of overlay capture.

    ``log`` accepts (level, message). It must be fast; use your logger's queue
    handler for slow destinations. ``context`` is an application-owned identity,
    never dereferenced by this module. No native objects are copied into history.

    ``on_changed()`` runs after mutations on the caller's thread; enqueue a
    coalesced UI redraw, never perform native rendering in the callback.
    ``context_key(context)`` maps the navigation capture to an opaque viewport
    identity. It must use that capture, not the newly active view. The default
    preserves the capture itself. Callback failures are isolated by ``observe``.
    Expiry alone emits no notification: schedule a refresh at ``expires_at``.
    """
    def __init__(self, *, enabled=False, log=None, clock=time.monotonic,
                 history_limit=30, retention=1.0, log_level="info",
                 on_changed=None, context_key=None):
        if history_limit < 1 or not math.isfinite(retention) or retention < 0:
            raise ValueError('Invalid diagnostic limits')
        import logging
        self.enabled, self.log, self.clock = enabled, log or (lambda level, message: getattr(logging.getLogger("openaxis"), level)(message)), clock
        self.retention = retention
        self.log_level = log_level
        self._logged_corrections = {}
        self._unknown_readback = {}
        self.history = deque(maxlen=history_limit)
        self.revision = 0
        self._comparison, self._object_comparison = compare, compare_object
        self.on_changed = None
        self.context_key = context_key or (lambda context: context)
        self.clear()
        self.on_changed = on_changed

    def _touch(self):
        self.revision += 1
        try:
            if self.on_changed is not None:
                self.on_changed()
        except Exception:
            pass  # Diagnostic redraw failures must not interrupt navigation.

    def bind(self, comparison, object_comparison=None):
        self._comparison, self._object_comparison = comparison, object_comparison or compare_object

    def clear(self):
        self._reset()
        self._touch()

    def _reset(self):
        self.query = None
        self.context = None
        self.writes = {}
        self.poses = {}
        self.corrections = {}
        self.status = None
        self.history.clear()

    def set_enabled(self, enabled):
        if self.enabled != enabled:
            self.enabled = enabled
            self.clear()

    def set_context(self, context):
        """Bind evidence to the viewport actually used by the query."""
        if self.enabled:
            self._set_context(context)
            self._touch()

    def _set_context(self, context):
        if self.enabled:
            if self.context != context:
                self.writes.clear()
                self.poses.clear()
                self.corrections.clear()
            self.context = context
            if self.query is not None and not self.query.complete:
                self.query.context = context

    def pick(self, request_id, name, *, screen=None, ray=None):
        """Attach actual pick evidence. Ray is (world start, world end)."""
        q = self.query
        if self.enabled and q and not q.complete and q.request_id == request_id:
            q.picks[name] = deepcopy((screen,ray))
            self._touch()

    def observe(self, event, **v):
        # Keep logger/observer errors from entering application navigation paths.
        try:
            self._observe(event, v)
        except Exception:
            # Diagnostics are best-effort. Native/session errors have their own
            # evidence; a broken diagnostic sink must not become one of them.
            pass

    def _observe(self, event, v):
        query = v.get('query')
        message = None
        level = 'info'
        if event == 'gesture_started':
            self._logged_corrections.clear()
            self._unknown_readback.clear()
        if event == 'fact':
            raw_value = v.get('value')
            value = _wire_fact_value(v['name'], raw_value)
            available = value is not None and value is not UNAVAILABLE
            message = format_event('navigation.fact',fact=v['name'],
                result='error' if v.get('error') else 'ok' if available else 'missing',
                value=value if available else None,error=v.get('error'),duration_ms=v['duration_ms'])
        elif event == 'query_completed':
            message = format_event('navigation.query.complete',request=query.request_id,
                first=(v['result'].get('first') or {}).get('name'),duration_ms=v['duration_ms'])
        elif event in ('gesture_started','gesture_finished','cancelled','output_rejected','query_failed'):
            level = 'warning' if event in ('output_rejected','query_failed') else 'info'
            message = f"{event}: " + ', '.join(f'{k}={x}' for k,x in v.items() if k != 'query')
        elif event in ('camera_write','object_write'):
            unknown = v['success'] and v.get('realized') is None
            was_unknown = self._unknown_readback.get(event, False)
            self._unknown_readback[event] = unknown
            level = 'warning' if not v['success'] else 'info' if unknown != was_unknown else 'debug'
            state = 'failed' if not v['success'] else 'unknown readback' if unknown else 'readback recovered' if was_unknown else 'succeeded'
            if not v['success'] or unknown != was_unknown:
                message = event.replace('_',' ') + ': ' + state
        elif 'correction_' in event:
            level = 'debug'
            stream = 'object' if event.startswith('object_') else 'camera'
            state = event.rsplit('_',1)[-1]
            key = (state,v['delta_id'])
            if self._logged_corrections.get(stream) != key:
                message = f"{stream} correction {v['delta_id']}: {'acknowledged' if state == 'applied' else state}"
                difference = v.get('difference')
                if difference is not None:
                    message += f' | translation {difference.t} | rotation {difference.r} rad | scale {difference.scale}'
                self._logged_corrections[stream] = key
        if message and self.log and (level != "debug" or self.log_level == "debug"):
            try:
                self.log(level,message)
            except Exception:
                pass
        if not self.enabled:
            return
        if message and event not in ('fact','camera_write','object_write'):
            self.history.append((self.clock(),level,message))
        if event == 'gesture_started':
            self._reset()
            self.status = f"gesture {v['gesture_id']} started"
        elif event in ('gesture_finished','cancelled'):
            self.status = f"gesture finished: {v.get('reason','motion_end')}"
            self.corrections = {k:(i,'ended',self.clock()+self.retention) for k,(i,_,_) in self.corrections.items()}
        elif event == 'query_started':
            self.query = QueryEvidence(query.request_id,query.gesture_id,tuple(query.values),tuple(query.first))
        elif event == 'query_context':
            self._set_context(self.context_key(v['context']))
        elif event == 'fact' and self.query and query.request_id == self.query.request_id:
            if v['name'].startswith('pick.') and isinstance(raw_value, dict):
                marker = raw_value.get('markerPosition')
                if isinstance(marker, (list, tuple)) and len(marker) == 2 and all(isinstance(x, (int, float)) and math.isfinite(x) for x in marker):
                    self.pick(query.request_id, v['name'], screen=tuple(marker))
            self.query.facts[v['name']] = (deepcopy(value) if available else None,v['duration_ms'],v.get('error'))
        elif event in ('query_completed','query_failed') and self.query:
            if query is not None and query.request_id != self.query.request_id:
                return
            self.query.complete = True
            self.query.duration_ms = v['duration_ms']
            self.query.error = v.get('error')
            self.query.selected = (v.get('result',{}).get('first') or {}).get('name')
        elif event in ('camera_write','object_write'):
            kind = event.split('_')[0]
            actual = v.get('realized')
            self.poses[kind] = deepcopy((v['desired'],actual))
            state = 'failed' if not v['success'] else 'unknown readback' if actual is None else 'comparison unavailable'
            comparison = self._comparison if kind == 'camera' else self._object_comparison
            detail = ''
            if v['success'] and actual is not None and comparison:
                difference = comparison(v['desired'],actual)
                state = 'differs' if difference.changed or difference.discontinuity else 'equivalent'
                detail = f' | translation {math.hypot(*difference.t):.3g} application units | rotation {math.degrees(math.hypot(*difference.r)):.3g} deg'
            self.writes[kind] = DiagnosticLine(f'{kind} write: {state}{detail}', 'missing' if state == 'failed' else 'pass' if state == 'equivalent' else 'correction')
        elif 'correction_' in event:
            level = 'debug'
            kind = 'object' if event.startswith('object_') else 'camera'
            state = event.rsplit('_',1)[-1]
            until = self.clock()+self.retention if state == 'applied' else None
            self.corrections[kind] = (v['delta_id'],'acknowledged' if state == 'applied' else state,until)
        elif event == 'output_rejected':
            self.status = message
        self._touch()

    def presentation(self):
        if not self.enabled:
            return DiagnosticPresentation(None,revision=self.revision)
        lines,segments,markers = [],[],[]
        q = self.query
        if q and q.context == self.context:
            lines.append(DiagnosticLine(f'Navigation diagnostics | gesture {q.gesture_id} | query {q.request_id} | {q.duration_ms:.1f} ms'))
            if q.error:
                lines.append(DiagnosticLine(q.error,'missing'))
            for name in dict.fromkeys(q.values+q.first):
                fact = q.facts.get(name)
                if fact is None:
                    state = 'skipped' if q.complete and q.selected and name in q.first and q.first.index(name)>q.first.index(q.selected) else 'not evaluated'
                    lines.append(DiagnosticLine(f'{name}: {state}','skipped'))
                else:
                    value,duration,error = fact
                    text = format_event('navigation.fact',fact=name,result='error' if error else 'ok' if value is not None else 'missing',value=value,error=error,duration_ms=duration)
                    lines.append(DiagnosticLine(text+(' < returned candidate' if name == q.selected else ''),'missing' if error or value is None else _tone(name)))
            scale = 1.
            for name in ('selection.bounds','model.bounds','object.bounds'):
                bounds = _bounds(q.facts.get(name,(None,))[0])
                if bounds:
                    scale = max(.0001,math.dist(*bounds)*.1)
                    break
            for name,(value,_,_) in q.facts.items():
                tone = _tone(name)
                if name == 'world.orientation' and isinstance(value,dict):
                    forward,up = _point(value.get('forward')),_point(value.get('up'))
                    if forward and up:
                        right = (forward[1]*up[2]-forward[2]*up[1],forward[2]*up[0]-forward[0]*up[2],forward[0]*up[1]-forward[1]*up[0])
                        sign = -1 if value.get('handedness') == 'left' else 1
                        for direction,axis_tone in ((tuple(sign*x for x in right),'axis_x'),(up,'axis_y'),(forward,'axis_z')):
                            segments.append(DiagnosticSegment((0,0,0),tuple(scale*x for x in direction),axis_tone,2))
                if name == 'object.pose' and isinstance(value,dict):
                    origin,rotation = _point(value.get('t')),_point(value.get('r'))
                    if origin and rotation:
                        qrot = Quat.from_rotvec(*rotation)
                        for axis,axis_tone in (((1,0,0),'axis_x'),((0,1,0),'axis_y'),((0,0,1),'axis_z')):
                            direction = qrot.rotate(Vec3(*axis))
                            end = tuple(a+scale*b for a,b in zip(origin,(direction.x,direction.y,direction.z)))
                            segments.append(DiagnosticSegment(origin,end,axis_tone,2))
                if name == 'sketch.plane' and isinstance(value,dict):
                    origin,normal,axis = (_point(value.get(k)) for k in ('origin','normal','x_axis'))
                    if origin and normal and axis and math.hypot(*normal)>0 and math.hypot(*axis)>0:
                        normal = tuple(v/math.hypot(*normal) for v in normal)
                        axis = tuple(v/math.hypot(*axis) for v in axis)
                        y = (normal[1]*axis[2]-normal[2]*axis[1],normal[2]*axis[0]-normal[0]*axis[2],normal[0]*axis[1]-normal[1]*axis[0])
                        if math.hypot(*y)>0:
                            y = tuple(v/math.hypot(*y) for v in y)
                            for step in (-1,-.5,0,.5,1):
                                for along,across in ((axis,y),(y,axis)):
                                    start = tuple(o+scale*(step*a-b) for o,a,b in zip(origin,along,across))
                                    end = tuple(o+scale*(step*a+b) for o,a,b in zip(origin,along,across))
                                    segments.append(DiagnosticSegment(start,end,'sketch',1))
                            segments.append(DiagnosticSegment(origin,tuple(o+scale*n for o,n in zip(origin,normal)),'target',2))
                bounds = _bounds(value)
                if bounds:
                    segments.extend(_box(bounds,tone))
                point = _point(value) if name in ('camera.view_target','scene.cursor') else _point(value.get('point')) if isinstance(value,dict) else None
                if point:
                    segments.extend(_cross(point,scale*.08,tone))
                if isinstance(value,dict) and (bounds := _bounds(value.get('bounds'))):
                    segments.extend(_box(bounds,tone))
            for name,(screen,ray) in q.picks.items():
                if screen is not None:
                    markers.append(DiagnosticMarker(name,tuple(screen),_tone(name)))
        if self.status:
            lines.append(DiagnosticLine(self.status))
        lines.extend(self.writes.values())
        expiries = []
        for kind,(identifier,state,until) in self.corrections.items():
            if until is None or self.clock() < until:
                lines.append(DiagnosticLine(f'{kind} correction {identifier}: {state}','correction'))
                if until is not None:
                    expiries.append(until)
        # Several queries can sample the same pixel. Preserve every query name
        # without drawing their crosshairs and labels on top of one another.
        grouped = {}
        for marker in markers:
            grouped.setdefault((marker.point,marker.tone),[]).append(marker.label)
        markers = [DiagnosticMarker('\n'.join(labels),point,tone)
                   for (point,tone),labels in grouped.items()]
        return DiagnosticPresentation(self.context,tuple(lines),tuple(segments),tuple(markers),self.revision,min(expiries) if expiries else None)

File: ts/sdk/src/navigation-diagnostics.ts
import { emitDiagnosticLog } from "./logging.js";
import { formatEvent, type DiagnosticLevel } from "./diagnostics.js";
import { Quat } from "./geometry/quat.js";
import { Vec3 } from "./geometry/vec3.js";
import { UNAVAILABLE, wireFactValue } from "./navigation.js";
import type { NavigationEvent, NavigationEventMap } from "./navigation-observer.js";
import { comparePoses, compareObjectPoses, type PoseComparison, type ObjectPoseComparison } from "./session-state.js";
import type { OpenAxisInteger, Vec3 as Point } from "./protocol/types.js";

export const DIAGNOSTIC_COLORS = Object.freeze({
  text: [245,245,245], missing: [255,130,130], skipped: [165,165,165], pass: [80,255,110],
  selection: [255,150,40], model: [40,210,255], target: [255,70,220], cursor: [255,235,40],
  center: [100,170,255], object: [255,70,220], sketch: [190,120,255], axis_x: [255,60,60],
  axis_y: [60,255,60], axis_z: [60,130,255], ray: [175,175,175], correction: [255,210,40],
} satisfies Record<string, readonly [number,number,number]>);
for (const color of Object.values(DIAGNOSTIC_COLORS)) Object.freeze(color);
export type DiagnosticTone = keyof typeof DIAGNOSTIC_COLORS;
export interface DiagnosticLine { text: string; tone: DiagnosticTone }
export interface DiagnosticSegment { start: Point; end: Point; tone: DiagnosticTone; width: number; opacity: number }
export interface DiagnosticMarker { label: string; point: [number,number]; tone: DiagnosticTone }
export interface DiagnosticPresentation {
  /** Opaque application identity; never dereferenced or cloned. */
  context: unknown;
  lines: DiagnosticLine[];
  segments: DiagnosticSegment[];
  markers: DiagnosticMarker[];
  revision: number;
  /** Monotonic seconds. Expiry alone does not emit onChanged. */
  expiresAt?: number;
}
export interface DiagnosticHistoryEntry { time: number; level: DiagnosticLevel; message: string }
export interface NavigationDiagnosticsOptions {
  enabled?: boolean;
  log?: (level: DiagnosticLevel, message: string) => void;
  clock?: () => number;
  historyLimit?: number;
  retention?: number;
  logLevel?: "info" | "debug";
  onChanged?: () => void;
  contextKey?: (context: unknown) => unknown;
}
interface Fact { value: unknown; durationMs: number; error?: string }
interface Pick { screen?: [number,number]; ray?: [Point,Point] }
interface QueryEvidence {
  requestId: OpenAxisInteger; gestureId?: OpenAxisInteger; values: string[]; first: string[];
  context?: unknown; facts: Map<string,Fact>; picks: Map<string,Pick>;
  selected?: string; durationMs: number; error?: string; complete: boolean;
}
type Stream = "camera" | "object";
const record = (v: unknown): Record<string,unknown> => v !== null && typeof v === "object" && !Array.isArray(v) ? v as Record<string,unknown> : {};
const point = (v: unknown): Point | undefined => Array.isArray(v) && v.length === 3 && v.every(x => typeof x === "number" && Number.isFinite(x)) ? [...v] as Point : undefined;
function bounds(v: unknown): [Point,Point] | undefined {
  const b = record(v), low = point(b.min), high = point(b.max);
  return low && high && low.every((x,i) => x <= high[i]!) ? [low,high] : undefined;
}
function tone(name: string): DiagnosticTone {
  if (name.startsWith("pick.")) return name.includes("cursor") ? "cursor" : "center";
  return ({ "selection.bounds": "selection", "model.bounds": "model", "object.bounds": "object", "object.pose": "object",
    "scene.cursor": "target", "sketch.plane": "sketch", "camera.view_target": "target" } as Record<string,DiagnosticTone>)[name] ?? "text";
}
const short = (n: number) => Number(n.toPrecision(3)).toString();

/** Passive, renderer-independent evidence. All operations belong to the session thread. */
export class NavigationDiagnostics {
  private active: boolean;
  private version = 0;
  private readonly clock: () => number;
  private readonly limit: number;
  private readonly retention: number;
  private entries: DiagnosticHistoryEntry[] = [];
  private query?: QueryEvidence;
  private context?: unknown;
  private status?: string;
  private writes = new Map<Stream,DiagnosticLine>();
  private corrections = new Map<Stream,{ id: OpenAxisInteger; state: string; until?: number }>();
  private loggedCorrections = new Map<Stream,string>();
  private unknownReadback = new Map<Stream,boolean>();
  private comparison: PoseComparison = comparePoses;
  private objectComparison: ObjectPoseComparison = compareObjectPoses;
  onChanged?: () => void;

  constructor(private readonly options: NavigationDiagnosticsOptions = {}) {
    this.active = options.enabled ?? false;
    this.clock = options.clock ?? (() => performance.now() / 1000);
    this.limit = options.historyLimit ?? 30;
    this.retention = options.retention ?? 1;
    if (!Number.isInteger(this.limit) || this.limit < 1 || !Number.isFinite(this.retention) || this.retention < 0) throw new TypeError("Invalid diagnostic limits");
    this.clear();
    this.onChanged = options.onChanged;
  }
  get enabled(): boolean { return this.active }
  get revision(): number { return this.version }
  get history(): DiagnosticHistoryEntry[] { return this.entries.map(x => ({ ...x })) }
  bind(comparison: PoseComparison = comparePoses, objectComparison: ObjectPoseComparison = compareObjectPoses): void {
    this.comparison = comparison; this.objectComparison = objectComparison;
  }
  private touch(): void { this.version++; try { this.onChanged?.() } catch { /* passive */ } }
  private reset(): void {
    this.query = undefined; this.context = undefined; this.status = undefined;
    this.writes.clear(); this.corrections.clear(); this.entries = [];
  }
  clear(): void { this.reset(); this.touch() }
  setEnabled(value: boolean): void { if (this.active !== value) { this.active = value; this.clear() } }
  private assignContext(context: unknown): void {
    if (this.context !== context) { this.writes.clear(); this.corrections.clear() }
    this.context = context;
    if (this.query && !this.query.complete) this.query.context = context;
  }
  setContext(context: unknown): void { if (this.active) { this.assignContext(context); this.touch() } }
  pick(requestId: OpenAxisInteger, name: string, evidence: Pick): void {
    if (this.active && this.query && !this.query.complete && this.query.requestId === requestId) {
      this.query.picks.set(name, structuredClone(evidence)); this.touch();
    }
  }
  observe<K extends keyof NavigationEventMap>(event: K, values: NavigationEventMap[K]): void {
    try { this.consume({ event, values } as NavigationEvent) } catch { /* A diagnostic failure cannot fail navigation. */ }
  }
  private log(level: DiagnosticLevel, message: string, retain = true): void {
    if (level !== "debug" || this.options.logLevel === "debug") {
      try { (this.options.log ?? emitDiagnosticLog)(level, message) } catch { /* passive */ }
    }
    if (this.active && retain) {
      this.entries.push({ time: this.clock(), level, message });
      if (this.entries.length > this.limit) this.entries.shift();
    }
  }
  private consume(e: NavigationEvent): void {
    switch (e.event) {
      case "gesture_started":
        this.loggedCorrections.clear(); this.unknownReadback.clear();
        this.log("info", `gesture_started: gesture_id=${e.values.gestureId}`, false);
        if (this.active) { this.reset(); this.status = `gesture ${e.values.gestureId} started` }
        break;
      case "gesture_finished": case "cancelled": {
        const { gestureId, reason } = e.values;
        this.log("info", `${e.event}: gesture_id=${gestureId}, reason=${reason}`);
        if (this.active) {
          this.status = `gesture finished: ${reason}`;
          for (const c of this.corrections.values()) { c.state = "ended"; c.until = this.clock() + this.retention }
        }
        break;
      }
      case "output_rejected": {
        const v = e.values, message = `output_rejected: kind=${v.kind}, gesture_id=${v.gestureId}, reason=${v.reason}`;
        this.log("warning", message); if (this.active) this.status = message;
        break;
      }
      case "query_started": {
        const q = e.values.query;
        if (this.active) this.query = { requestId: q.requestId, gestureId: q.gestureId, values: [...q.values], first: [...q.first],
          facts: new Map(), picks: new Map(), durationMs: 0, complete: false };
        break;
      }
      case "query_context":
        if (this.active) this.assignContext(this.options.contextKey ? this.options.contextKey(e.values.context) : e.values.context);
        break;
      case "fact": {
        const v = e.values, value = wireFactValue(v.name, v.value), available = value != null && value !== UNAVAILABLE;
        const marker = record(v.value).markerPosition;
        if (v.name.startsWith("pick.") && Array.isArray(marker) && marker.length === 2 && marker.every(x => typeof x === "number" && Number.isFinite(x)))
          this.pick(v.query.requestId, v.name, { screen: [marker[0], marker[1]] });
        this.log("info", formatEvent("navigation.fact", { fact: v.name, result: v.error ? "error" : available ? "ok" : "missing",
          value: available ? value : undefined, error: v.error, duration_ms: v.durationMs }), false);
        if (this.active && this.query?.requestId === v.query.requestId) this.query.facts.set(v.name,
          { value: available ? structuredClone(value) : undefined, durationMs: v.durationMs, error: v.error });
        break;
      }
      case "query_completed": case "query_failed": {
        const v = e.values, selected = e.event === "query_completed" ? record(e.values.result.first).name as string | undefined : undefined;
        const error = e.event === "query_failed" ? e.values.error : undefined;
        this.log(error ? "warning" : "info", error ? `query_failed: error=${error}, duration_ms=${v.durationMs}`
          : formatEvent("navigation.query.complete", { request: v.query.requestId, first: selected, duration_ms: v.durationMs }));
        if (this.active && this.query?.requestId === v.query.requestId) {
          Object.assign(this.query, { complete: true, durationMs: v.durationMs, selected, error });
        }
        break;
      }
      case "camera_write": case "object_write": {
        const stream = e.event === "camera_write" ? "camera" : "object", v = e.values;
        const unknown = v.success && !v.realized, previous = this.unknownReadback.get(stream) ?? false;
        this.unknownReadback.set(stream, unknown);
        const level = !v.success ? "warning" : unknown !== previous ? "info" : "debug";
        if (!v.success || unknown !== previous) this.log(level, `${stream} write: ${!v.success ? "failed" : unknown ? "unknown readback" : previous ? "readback recovered" : "succeeded"}`, false);
        if (this.active) {
          let state = !v.success ? "failed" : !v.realized ? "unknown readback" : "comparison unavailable", detail = "";
          if (v.success && v.realized) {
            const d = (stream === "camera" ? this.comparison : this.objectComparison)(v.desired, v.realized);
            state = d.changed || d.discontinuity ? "differs" : "equivalent";
            detail = ` | translation ${short(Math.hypot(...d.t))} application units | rotation ${short(Math.hypot(...d.r) * 180 / Math.PI)} deg`;
          }
          this.writes.set(stream, { text: `${stream} write: ${state}${detail}`, tone: state === "failed" ? "missing" : state === "equivalent" ? "pass" : "correction" });
        }
        break;
      }
      case "correction_sent": case "correction_waiting": case "correction_applied":
      case "object_correction_sent": case "object_correction_waiting": case "object_correction_applied": {
        const stream = e.event.startsWith("object_") ? "object" : "camera", v = e.values;
        const state = e.event.endsWith("applied") ? "acknowledged" : e.event.endsWith("waiting") ? "waiting" : "sent";
        const key = `${state}:${v.deltaId}`;
        if (this.loggedCorrections.get(stream) !== key) {
          const d = v.difference;
          this.log("debug", `${stream} correction ${v.deltaId}: ${state}` + (d ? ` | translation ${JSON.stringify(d.t)} | rotation ${JSON.stringify(d.r)} rad | scale ${d.scale}` : ""));
          this.loggedCorrections.set(stream, key);
        }
        if (this.active) this.corrections.set(stream, { id: v.deltaId, state, until: state === "acknowledged" ? this.clock() + this.retention : undefined });
        break;
      }
      default: return;
    }
    if (this.active) this.touch();
  }

  presentation(): DiagnosticPresentation {
    const frame: DiagnosticPresentation = { context: this.active ? this.context : undefined, lines: [], segments: [], markers: [], revision: this.version };
    if (!this.active) return frame;
    const { lines, segments, markers } = frame, q = this.query;
    const line = (text: string, tone: DiagnosticTone = "text") => lines.push({ text, tone });
    const segment = (start: Point, end: Point, tone: DiagnosticTone, width = 2, opacity = 1) => segments.push({ start: [...start], end: [...end], tone, width, opacity });
    if (q && q.context === this.context) {
      line(`Navigation diagnostics | gesture ${q.gestureId ?? "None"} | query ${q.requestId} | ${q.durationMs.toFixed(1)} ms`);
      if (q.error) line(q.error, "missing");
      for (const name of new Set([...q.values,...q.first])) {
        const f = q.facts.get(name);
        if (!f) line(`${name}: ${q.complete && q.selected && q.first.indexOf(name) > q.first.indexOf(q.selected) ? "skipped" : "not evaluated"}`, "skipped");
        else line(formatEvent("navigation.fact", { fact: name, result: f.error ? "error" : f.value == null ? "missing" : "ok",
          value: f.value, error: f.error, duration_ms: f.durationMs }) + (name === q.selected ? " < returned candidate" : ""), f.error || f.value == null ? "missing" : tone(name));
      }
      let scale = 1;
      for (const name of ["selection.bounds","model.bounds","object.bounds"]) {
        const b = bounds(q.facts.get(name)?.value);
        if (b) { scale = Math.max(.0001, Vec3.fromArray(b[1]).sub(Vec3.fromArray(b[0])).length() * .1); break }
      }
      const box = (b: [Point,Point], color: DiagnosticTone) => {
        const corners: Point[] = [];
        for (const x of [b[0][0],b[1][0]]) for (const y of [b[0][1],b[1][1]]) for (const z of [b[0][2],b[1][2]]) corners.push([x,y,z]);
        corners.forEach((start,i) => { for (const bit of [1,2,4]) if (!(i & bit)) segment(start,corners[i | bit]!,color,1,.35) });
      };
      for (const [name,f] of q.facts) {
        const v = record(f.value), color = tone(name);
        if (name === "world.orientation") {
          const forward = point(v.forward), up = point(v.up);
          if (forward && up) {
            const right = Vec3.fromArray(forward).cross(Vec3.fromArray(up)).mul(v.handedness === "left" ? -1 : 1).toArray();
            for (const [direction,color] of [[right,"axis_x"],[up,"axis_y"],[forward,"axis_z"]] as const) segment([0,0,0],Vec3.fromArray(direction).mul(scale).toArray(),color);
          }
        }
        if (name === "object.pose") {
          const origin = point(v.t), rotation = point(v.r);
          if (origin && rotation) {
            const quat = Quat.fromRotvec(...rotation);
            for (const [axis,color] of [[[1,0,0],"axis_x"],[[0,1,0],"axis_y"],[[0,0,1],"axis_z"]] as const)
              segment(origin,Vec3.fromArray(origin).add(quat.rotate(Vec3.fromArray(axis)).mul(scale)).toArray(),color);
          }
        }
        if (name === "sketch.plane") {
          const origin = point(v.origin), normal = point(v.normal), axis = point(v.x_axis);
          if (origin && normal && axis && Math.hypot(...normal) > 0 && Math.hypot(...axis) > 0) {
            const o = Vec3.fromArray(origin), n = Vec3.fromArray(normal).normalized(), x = Vec3.fromArray(axis).normalized(), y = n.cross(x).normalized();
            if (y.length() > 0) {
              for (const step of [-1,-.5,0,.5,1]) for (const [along,across] of [[x,y],[y,x]] as const)
                segment(o.add(along.mul(step * scale)).sub(across.mul(scale)).toArray(),o.add(along.mul(step * scale)).add(across.mul(scale)).toArray(),"sketch",1);
              segment(origin,o.add(n.mul(scale)).toArray(),"target");
            }
          }
        }
        const b = bounds(f.value); if (b) box(b,color);
        const p = name === "camera.view_target" || name === "scene.cursor" ? point(f.value) : point(v.point);
        if (p) for (let axis = 0; axis < 3; axis++) {
          const a: Point = [...p], b: Point = [...p]; a[axis]! -= scale * .08; b[axis]! += scale * .08; segment(a,b,color,1,.35);
        }
        const nested = bounds(v.bounds); if (nested) box(nested,color);
      }
      for (const [name,pick] of q.picks) {
        if (pick.screen) {
          const p = pick.screen, color = tone(name);
          const match = markers.find(m => m.tone === color && m.point[0] === p[0] && m.point[1] === p[1]);
          if (match) match.label += `\n${name}`; else markers.push({ label: name, point: [...p], tone: color });
        }
      }
    }
    if (this.status) line(this.status);
    for (const write of this.writes.values()) lines.push({ ...write });
    const now = this.clock();
    for (const [kind,c] of this.corrections) if (c.until === undefined || now < c.until) {
      line(`${kind} correction ${c.id}: ${c.state}`,"correction");
      if (c.until !== undefined) frame.expiresAt = Math.min(frame.expiresAt ?? Infinity,c.until);
    }
    return frame;
  }
}
